Only in .: 1.1.4.patch
diff -cwr ./adler32.c d:\djgpp\contrib\zlib-1.1.4/adler32.c
*** ./adler32.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/adler32.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* adler32.c -- compute the Adler-32 checksum of a data stream
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* adler32.c -- compute the Adler-32 checksum of a data stream
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: adler32.d
Only in d:\djgpp\contrib\zlib-1.1.4: adler32.o
Only in d:\djgpp\contrib\zlib-1.1.4: amiga
diff -cwr ./compress.c d:\djgpp\contrib\zlib-1.1.4/compress.c
*** ./compress.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/compress.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* compress.c -- compress a memory buffer
!  * Copyright (C) 1995-1998 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* compress.c -- compress a memory buffer
!  * Copyright (C) 1995-2002 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: compress.d
Only in d:\djgpp\contrib\zlib-1.1.4: compress.o
Only in d:\djgpp\contrib\zlib-1.1.4: contrib
Only in d:\djgpp\contrib\zlib-1.1.4: crc32.d
Only in d:\djgpp\contrib\zlib-1.1.4: crc32.o
Only in .: Debug
diff -cwr ./deflate.c d:\djgpp\contrib\zlib-1.1.4/deflate.c
*** ./deflate.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/deflate.c	Mon Mar 11 10:00:38 2002
***************
*** 1,5 ****
  /* deflate.c -- compress data using the deflation algorithm
!  * Copyright (C) 1995-1998 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* deflate.c -- compress data using the deflation algorithm
!  * Copyright (C) 1995-2002 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
***************
*** 52,58 ****
  #include "deflate.h"
  
  const char deflate_copyright[] =
!    " deflate 1.1.3 Copyright 1995-1998 Jean-loup Gailly ";
  /*
    If you use the zlib library in a product, an acknowledgment is welcome
    in the documentation of your product. If for some reason you cannot
--- 52,58 ----
  #include "deflate.h"
  
  const char deflate_copyright[] =
!    " deflate 1.1.4 Copyright 1995-2002 Jean-loup Gailly ";
  /*
    If you use the zlib library in a product, an acknowledgment is welcome
    in the documentation of your product. If for some reason you cannot
***************
*** 242,248 ****
          windowBits = -windowBits;
      }
      if (memLevel < 1 || memLevel > MAX_MEM_LEVEL || method != Z_DEFLATED ||
!         windowBits < 8 || windowBits > 15 || level < 0 || level > 9 ||
  	strategy < 0 || strategy > Z_HUFFMAN_ONLY) {
          return Z_STREAM_ERROR;
      }
--- 242,248 ----
          windowBits = -windowBits;
      }
      if (memLevel < 1 || memLevel > MAX_MEM_LEVEL || method != Z_DEFLATED ||
!         windowBits < 9 || windowBits > 15 || level < 0 || level > 9 ||
  	strategy < 0 || strategy > Z_HUFFMAN_ONLY) {
          return Z_STREAM_ERROR;
      }
Only in d:\djgpp\contrib\zlib-1.1.4: deflate.d
diff -cwr ./deflate.h d:\djgpp\contrib\zlib-1.1.4/deflate.h
*** ./deflate.h	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/deflate.h	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* deflate.h -- internal compression state
!  * Copyright (C) 1995-1998 Jean-loup Gailly
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* deflate.h -- internal compression state
!  * Copyright (C) 1995-2002 Jean-loup Gailly
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: deflate.o
Only in d:\djgpp\contrib\zlib-1.1.4: example.c
Only in d:\djgpp\contrib\zlib-1.1.4: example.d
Only in d:\djgpp\contrib\zlib-1.1.4: example.exe
diff -cwr ./FAQ d:\djgpp\contrib\zlib-1.1.4/FAQ
*** ./FAQ	Sat Jul 08 17:51:40 2000
--- d:\djgpp\contrib\zlib-1.1.4/FAQ	Mon Mar 11 09:15:28 2002
***************
*** 3,72 ****
  
  
  If your question is not there, please check the zlib home page 
! http://www.cdrom.com/pub/infozip/zlib/ which may have more recent information.
  
  
! 1) I need a Windows DLL
! 2) I need a Visual Basic interface to zlib
! 3) compress() returns Z_BUF_ERROR
! 4) deflate or inflate returns Z_BUF_ERROR
! 5) Where is the zlib documentation (man pages, etc...)?
! 6) Why don't you use GNU autoconf, libtool, etc...?
! 7) There is a bug in zlib.
! 8) I get "undefined reference to gzputc"
  
  
  
! 1) I need a Windows DLL
  
!   The zlib sources can be compiled without change to produce a DLL.
!   If you want a precompiled DLL, see http://www.winimage.com/zLibDll
  
  
! 2) I need a Visual Basic interface to zlib
  
!   See http://www.tcfb.com/dowseware/cmp-z-it.zip
!       http://web2.airmail.net/markn/articles/zlibtool/zlibtool.htm
!   and contrib/visual-basic.txt
  
! 3) compress() returns Z_BUF_ERROR
  
!   Make sure that before the call of compress, the length of the
!   compressed buffer is equal to the total size of the compressed buffer
!   and not zero.  For Visual Basic, check that this parameter is passed
!   by reference ("as any"), not by value ("as long").
  
  
! 4) deflate or inflate returns Z_BUF_ERROR
  
!   Make sure that before the call avail_in and avail_out are not zero.
  
  
! 5) Where is the zlib documentation (man pages, etc...)?
  
!   It's in zlib.h for the moment. Volunteers to transform this
!   to man pages, please contact jloup@gzip.org. Examples of zlib usage
!   are in the files example.c and minigzip.c.
  
  
! 6) Why don't you use GNU autoconf, libtool, etc...?
  
!   Because we would like to keep zlib as a very small and simple package.
!   zlib is rather portable and doesn't need much configuration.
  
  
! 7) There is a bug in zlib.
  
!   Most of the time, such problems are due to an incorrect usage
!   of zlib. Please try to reproduce the problem with a small
!   program and send us the corresponding source at zlib@quest.jpl.nasa.gov
!   Do not send multi-megabyte data files without prior agreement.
  
  
! 8) I get "undefined reference to gzputc"
  
!   If "make test" produces something like
!      example.o(.text+0x174): 
!   check that you don't have old files libz.* in /usr/lib, /usr/local/lib
!   or /usr/X11R6/lib. Remove old versions then do "make install".
  
--- 3,100 ----
  
  
  If your question is not there, please check the zlib home page 
! http://www.zlib.org which may have more recent information.
! The lastest zlib FAQ is at http://www.gzip.org/zlib/zlib_faq.html
  
  
!  1. Is zlib Y2K-compliant?
  
+     Yes. zlib doesn't handle dates.
  
+  2. Where can I get a Windows DLL version?
  
!     The zlib sources can be compiled without change to produce a DLL. If you
!     want a precompiled DLL, see http://www.winimage.com/zLibDll/ . Questions
!     about the zlib DLL should be sent to Gilles Vollant (info@winimage.com).
  
!  3. Where can I get a Visual Basic interface to zlib?
  
+     See
+         * http://www.winimage.com/zLibDll/cmp-z-it.zip
+         * http://www.dogma.net/markn/articles/zlibtool/zlibtool.htm
+         * contrib/visual-basic.txt in the zlib distribution
  
!  4. compress() returns Z_BUF_ERROR
  
!     Make sure that before the call of compress, the length of the compressed
!     buffer is equal to the total size of the compressed buffer and not
!     zero. For Visual Basic, check that this parameter is passed by reference
!     ("as any"), not by value ("as long").
  
!  5. deflate() or inflate() returns Z_BUF_ERROR
  
!     Before making the call, make sure that avail_in and avail_out are not
!     zero. When setting the parameter flush equal to Z_FINISH, also make sure
!     that avail_out is big enough to allow processing all pending input.
  
+  6. Where's the zlib documentation (man pages, etc.)?
  
!     It's in zlib.h for the moment, and Francis S. Lin has converted it to a
!     web page zlib.html. Volunteers to transform this to Unix-style man pages,
!     please contact Jean-loup Gailly (jloup@gzip.org). Examples of zlib usage
!     are in the files example.c and minigzip.c.
  
!  7. Why don't you use GNU autoconf or libtool or ...?
  
+     Because we would like to keep zlib as a very small and simple
+     package. zlib is rather portable and doesn't need much configuration.
  
!  8. I found a bug in zlib.
  
!     Most of the time, such problems are due to an incorrect usage of
!     zlib. Please try to reproduce the problem with a small program and send
!     the corresponding source to us at zlib@gzip.org . Do not send
!     multi-megabyte data files without prior agreement.
  
+  9. Why do I get "undefined reference to gzputc"?
  
!     If "make test" produces something like
  
!        example.o(.text+0x154): undefined reference to `gzputc'
        
+     check that you don't have old files libz.* in /usr/lib, /usr/local/lib or
+     /usr/X11R6/lib. Remove any old versions, then do "make install".
  
! 10. I need a Delphi interface to zlib.
  
!     See the directories contrib/delphi and contrib/delphi2 in the zlib
!     distribution.
  
+ 11. Can zlib handle .zip archives?
  
!     See the directory contrib/minizip in the zlib distribution.
  
! 12. Can zlib handle .Z files?
! 
!     No, sorry. You have to spawn an uncompress or gunzip subprocess, or adapt
!     the code of uncompress on your own.
! 
! 13. How can I make a Unix shared library?
! 
!     make clean
!     ./configure -s
!     make
! 
! 14. Why does "make test" fail on Mac OS X?
! 
!     Mac OS X already includes zlib as a shared library, and so -lz links the
!     shared library instead of the one that the "make" compiled. For zlib
!     1.1.3, the two are incompatible due to different compile-time
!     options. Simply change the -lz in the Makefile to libz.a, and it will use
!     the compiled library instead of the shared one and the "make test" will
!     succeed.
! 
! 15. I have a question about OttoPDF
  
+     We are not the authors of OttoPDF. The real author is on the OttoPDF web
+     site Joel Hainley jhainley@myndkryme.com.
diff -cwr ./gzio.c d:\djgpp\contrib\zlib-1.1.4/gzio.c
*** ./gzio.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/gzio.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* gzio.c -- IO on .gz files
!  * Copyright (C) 1995-1998 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h
   *
   * Compile this file with -DNO_DEFLATE to avoid the compression code.
--- 1,5 ----
  /* gzio.c -- IO on .gz files
!  * Copyright (C) 1995-2002 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h
   *
   * Compile this file with -DNO_DEFLATE to avoid the compression code.
Only in d:\djgpp\contrib\zlib-1.1.4: gzio.d
Only in d:\djgpp\contrib\zlib-1.1.4: gzio.o
diff -cwr ./infblock.c d:\djgpp\contrib\zlib-1.1.4/infblock.c
*** ./infblock.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/infblock.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* infblock.c -- interpret and process block types to last block
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* infblock.c -- interpret and process block types to last block
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
***************
*** 249,258 ****
                               &s->sub.trees.tb, s->hufts, z);
        if (t != Z_OK)
        {
-         ZFREE(z, s->sub.trees.blens);
          r = t;
          if (r == Z_DATA_ERROR)
            s->mode = BAD;
          LEAVE
        }
        s->sub.trees.index = 0;
--- 249,260 ----
                               &s->sub.trees.tb, s->hufts, z);
        if (t != Z_OK)
        {
          r = t;
          if (r == Z_DATA_ERROR)
+         {
+           ZFREE(z, s->sub.trees.blens);
            s->mode = BAD;
+         }
          LEAVE
        }
        s->sub.trees.index = 0;
***************
*** 313,323 ****
          t = inflate_trees_dynamic(257 + (t & 0x1f), 1 + ((t >> 5) & 0x1f),
                                    s->sub.trees.blens, &bl, &bd, &tl, &td,
                                    s->hufts, z);
-         ZFREE(z, s->sub.trees.blens);
          if (t != Z_OK)
          {
            if (t == (uInt)Z_DATA_ERROR)
              s->mode = BAD;
            r = t;
            LEAVE
          }
--- 315,327 ----
          t = inflate_trees_dynamic(257 + (t & 0x1f), 1 + ((t >> 5) & 0x1f),
                                    s->sub.trees.blens, &bl, &bd, &tl, &td,
                                    s->hufts, z);
          if (t != Z_OK)
          {
            if (t == (uInt)Z_DATA_ERROR)
+           {
+             ZFREE(z, s->sub.trees.blens);
              s->mode = BAD;
+           }
            r = t;
            LEAVE
          }
***************
*** 329,334 ****
--- 333,339 ----
          }
          s->sub.decode.codes = c;
        }
+       ZFREE(z, s->sub.trees.blens);
        s->mode = CODES;
      case CODES:
        UPDATE
Only in d:\djgpp\contrib\zlib-1.1.4: infblock.d
diff -cwr ./infblock.h d:\djgpp\contrib\zlib-1.1.4/infblock.h
*** ./infblock.h	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/infblock.h	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* infblock.h -- header to use infblock.c
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* infblock.h -- header to use infblock.c
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: infblock.o
diff -cwr ./infcodes.c d:\djgpp\contrib\zlib-1.1.4/infcodes.c
*** ./infcodes.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/infcodes.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* infcodes.c -- process literals and length/distance pairs
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* infcodes.c -- process literals and length/distance pairs
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
***************
*** 196,210 ****
        Tracevv((stderr, "inflate:         distance %u\n", c->sub.copy.dist));
        c->mode = COPY;
      case COPY:          /* o: copying bytes in window, waiting for space */
- #ifndef __TURBOC__ /* Turbo C bug for following expression */
-       f = (uInt)(q - s->window) < c->sub.copy.dist ?
-           s->end - (c->sub.copy.dist - (q - s->window)) :
-           q - c->sub.copy.dist;
- #else
        f = q - c->sub.copy.dist;
!       if ((uInt)(q - s->window) < c->sub.copy.dist)
!         f = s->end - (c->sub.copy.dist - (uInt)(q - s->window));
! #endif
        while (c->len)
        {
          NEEDOUT
--- 196,204 ----
        Tracevv((stderr, "inflate:         distance %u\n", c->sub.copy.dist));
        c->mode = COPY;
      case COPY:          /* o: copying bytes in window, waiting for space */
        f = q - c->sub.copy.dist;
!       while (f < s->window)             /* modulo window size-"while" instead */
!         f += s->end - s->window;        /* of "if" handles invalid distances */
        while (c->len)
        {
          NEEDOUT
Only in d:\djgpp\contrib\zlib-1.1.4: infcodes.d
diff -cwr ./infcodes.h d:\djgpp\contrib\zlib-1.1.4/infcodes.h
*** ./infcodes.h	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/infcodes.h	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* infcodes.h -- header to use infcodes.c
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* infcodes.h -- header to use infcodes.c
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: infcodes.o
diff -cwr ./inffast.c d:\djgpp\contrib\zlib-1.1.4/inffast.c
*** ./inffast.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/inffast.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* inffast.c -- process literals and length/distance pairs fast
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* inffast.c -- process literals and length/distance pairs fast
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
***************
*** 93,120 ****
  
              /* do the copy */
              m -= c;
-             if ((uInt)(q - s->window) >= d)     /* offset before dest */
-             {                                   /*  just copy */
                r = q - d;
!               *q++ = *r++;  c--;        /* minimum count is three, */
!               *q++ = *r++;  c--;        /*  so unroll loop a little */
!             }
!             else                        /* else offset after destination */
              {
!               e = d - (uInt)(q - s->window); /* bytes from offset to end */
!               r = s->end - e;           /* pointer to offset */
!               if (c > e)                /* if source crosses, */
                {
!                 c -= e;                 /* copy to end of window */
                  do {
                    *q++ = *r++;
                  } while (--e);
!                 r = s->window;          /* copy rest from start of window */
                }
              }
!             do {                        /* copy all or what's left */
                *q++ = *r++;
              } while (--c);
              break;
            }
            else if ((e & 64) == 0)
--- 93,133 ----
  
              /* do the copy */
              m -= c;
              r = q - d;
!             if (r < s->window)                  /* wrap if needed */
              {
!               do {
!                 r += s->end - s->window;        /* force pointer in window */
!               } while (r < s->window);          /* covers invalid distances */
!               e = s->end - r;
!               if (c > e)
                {
!                 c -= e;                         /* wrapped copy */
                  do {
                      *q++ = *r++;
                  } while (--e);
!                 r = s->window;
!                 do {
!                     *q++ = *r++;
!                 } while (--c);
!               }
!               else                              /* normal copy */
!               {
!                 *q++ = *r++;  c--;
!                 *q++ = *r++;  c--;
!                 do {
!                     *q++ = *r++;
!                 } while (--c);
                }
              }
!             else                                /* normal copy */
!             {
!               *q++ = *r++;  c--;
!               *q++ = *r++;  c--;
!               do {
                  *q++ = *r++;
                } while (--c);
+             }
              break;
            }
            else if ((e & 64) == 0)
Only in d:\djgpp\contrib\zlib-1.1.4: inffast.d
diff -cwr ./inffast.h d:\djgpp\contrib\zlib-1.1.4/inffast.h
*** ./inffast.h	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/inffast.h	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* inffast.h -- header to use inffast.c
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* inffast.h -- header to use inffast.c
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: inffast.o
diff -cwr ./inflate.c d:\djgpp\contrib\zlib-1.1.4/inflate.c
*** ./inflate.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/inflate.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* inflate.c -- zlib interface to inflate modules
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* inflate.c -- zlib interface to inflate modules
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: inflate.d
Only in d:\djgpp\contrib\zlib-1.1.4: inflate.o
diff -cwr ./inftrees.c d:\djgpp\contrib\zlib-1.1.4/inftrees.c
*** ./inftrees.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/inftrees.c	Mon Mar 11 08:18:06 2002
***************
*** 1,5 ****
  /* inftrees.c -- generate Huffman trees for efficient decoding
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* inftrees.c -- generate Huffman trees for efficient decoding
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
***************
*** 11,17 ****
  #endif
  
  const char inflate_copyright[] =
!    " inflate 1.1.3 Copyright 1995-1998 Mark Adler ";
  /*
    If you use the zlib library in a product, an acknowledgment is welcome
    in the documentation of your product. If for some reason you cannot
--- 11,17 ----
  #endif
  
  const char inflate_copyright[] =
!    " inflate 1.1.4 Copyright 1995-2002 Mark Adler ";
  /*
    If you use the zlib library in a product, an acknowledgment is welcome
    in the documentation of your product. If for some reason you cannot
***************
*** 104,111 ****
  /* Given a list of code lengths and a maximum table size, make a set of
     tables to decode that set of codes.  Return Z_OK on success, Z_BUF_ERROR
     if the given code set is incomplete (the tables are still built in this
!    case), Z_DATA_ERROR if the input is invalid (an over-subscribed set of
!    lengths), or Z_MEM_ERROR if not enough memory. */
  {
  
    uInt a;                       /* counter for codes of length k */
--- 104,110 ----
  /* Given a list of code lengths and a maximum table size, make a set of
     tables to decode that set of codes.  Return Z_OK on success, Z_BUF_ERROR
     if the given code set is incomplete (the tables are still built in this
!    case), or Z_DATA_ERROR if the input is invalid. */
  {
  
    uInt a;                       /* counter for codes of length k */
***************
*** 231,237 ****
  
          /* allocate new table */
          if (*hn + z > MANY)     /* (note: doesn't matter for fixed) */
!           return Z_MEM_ERROR;   /* not enough memory */
          u[h] = q = hp + *hn;
          *hn += z;
  
--- 230,236 ----
  
          /* allocate new table */
          if (*hn + z > MANY)     /* (note: doesn't matter for fixed) */
!           return Z_DATA_ERROR;  /* overflow of MANY */
          u[h] = q = hp + *hn;
          *hn += z;
  
Only in d:\djgpp\contrib\zlib-1.1.4: inftrees.d
diff -cwr ./inftrees.h d:\djgpp\contrib\zlib-1.1.4/inftrees.h
*** ./inftrees.h	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/inftrees.h	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* inftrees.h -- header to use inftrees.c
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* inftrees.h -- header to use inftrees.c
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: inftrees.o
diff -cwr ./infutil.c d:\djgpp\contrib\zlib-1.1.4/infutil.c
*** ./infutil.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/infutil.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* inflate_util.c -- data and routines common to blocks and codes
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* inflate_util.c -- data and routines common to blocks and codes
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: infutil.d
diff -cwr ./infutil.h d:\djgpp\contrib\zlib-1.1.4/infutil.h
*** ./infutil.h	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/infutil.h	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* infutil.h -- types and macros common to blocks and codes
!  * Copyright (C) 1995-1998 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* infutil.h -- types and macros common to blocks and codes
!  * Copyright (C) 1995-2002 Mark Adler
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: infutil.o
Only in d:\djgpp\contrib\zlib-1.1.4: Make_vms.com
diff -cwr ./Makefile d:\djgpp\contrib\zlib-1.1.4/Makefile
*** ./Makefile	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/Makefile	Mon Mar 11 08:58:30 2002
***************
*** 1,5 ****
  # Makefile for zlib
! # Copyright (C) 1995-1998 Jean-loup Gailly.
  # For conditions of distribution and use, see copyright notice in zlib.h 
  
  # To compile and test, type:
--- 1,5 ----
  # Makefile for zlib
! # Copyright (C) 1995-2002 Jean-loup Gailly.
  # For conditions of distribution and use, see copyright notice in zlib.h 
  
  # To compile and test, type:
***************
*** 24,30 ****
  LDSHARED=$(CC)
  CPP=$(CC) -E
  
! VER=1.1.3
  LIBS=libz.a
  SHAREDLIB=libz.so
  
--- 24,30 ----
  LDSHARED=$(CC)
  CPP=$(CC) -E
  
! VER=1.1.4
  LIBS=libz.a
  SHAREDLIB=libz.so
  
***************
*** 47,53 ****
  TEST_OBJS = example.o minigzip.o
  
  DISTFILES = README FAQ INDEX ChangeLog configure Make*[a-z0-9] *.[ch] *.mms \
!   algorithm.txt zlib.3 msdos/Make*[a-z0-9] msdos/zlib.def msdos/zlib.rc \
    nt/Make*[a-z0-9] nt/zlib.dnt amiga/Make*.??? os2/M*.os2 os2/zlib.def \
    contrib/RE*.contrib contrib/*.txt contrib/asm386/*.asm contrib/asm386/*.c \
    contrib/asm386/*.bat contrib/asm386/zlibvc.d?? contrib/asm[56]86/*.?86 \
--- 47,54 ----
  TEST_OBJS = example.o minigzip.o
  
  DISTFILES = README FAQ INDEX ChangeLog configure Make*[a-z0-9] *.[ch] *.mms \
!   algorithm.txt zlib.3 zlib.html \
!   msdos/Make*[a-z0-9] msdos/zlib.def msdos/zlib.rc \
    nt/Make*[a-z0-9] nt/zlib.dnt amiga/Make*.??? os2/M*.os2 os2/zlib.def \
    contrib/RE*.contrib contrib/*.txt contrib/asm386/*.asm contrib/asm386/*.c \
    contrib/asm386/*.bat contrib/asm386/zlibvc.d?? contrib/asm[56]86/*.?86 \
diff -cwr ./Makefile.in d:\djgpp\contrib\zlib-1.1.4/Makefile.in
*** ./Makefile.in	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/Makefile.in	Mon Mar 11 08:58:30 2002
***************
*** 1,5 ****
  # Makefile for zlib
! # Copyright (C) 1995-1998 Jean-loup Gailly.
  # For conditions of distribution and use, see copyright notice in zlib.h 
  
  # To compile and test, type:
--- 1,5 ----
  # Makefile for zlib
! # Copyright (C) 1995-2002 Jean-loup Gailly.
  # For conditions of distribution and use, see copyright notice in zlib.h 
  
  # To compile and test, type:
***************
*** 24,30 ****
  LDSHARED=$(CC)
  CPP=$(CC) -E
  
! VER=1.1.3
  LIBS=libz.a
  SHAREDLIB=libz.so
  
--- 24,30 ----
  LDSHARED=$(CC)
  CPP=$(CC) -E
  
! VER=1.1.4
  LIBS=libz.a
  SHAREDLIB=libz.so
  
***************
*** 47,53 ****
  TEST_OBJS = example.o minigzip.o
  
  DISTFILES = README FAQ INDEX ChangeLog configure Make*[a-z0-9] *.[ch] *.mms \
!   algorithm.txt zlib.3 msdos/Make*[a-z0-9] msdos/zlib.def msdos/zlib.rc \
    nt/Make*[a-z0-9] nt/zlib.dnt amiga/Make*.??? os2/M*.os2 os2/zlib.def \
    contrib/RE*.contrib contrib/*.txt contrib/asm386/*.asm contrib/asm386/*.c \
    contrib/asm386/*.bat contrib/asm386/zlibvc.d?? contrib/asm[56]86/*.?86 \
--- 47,54 ----
  TEST_OBJS = example.o minigzip.o
  
  DISTFILES = README FAQ INDEX ChangeLog configure Make*[a-z0-9] *.[ch] *.mms \
!   algorithm.txt zlib.3 zlib.html \
!   msdos/Make*[a-z0-9] msdos/zlib.def msdos/zlib.rc \
    nt/Make*[a-z0-9] nt/zlib.dnt amiga/Make*.??? os2/M*.os2 os2/zlib.def \
    contrib/RE*.contrib contrib/*.txt contrib/asm386/*.asm contrib/asm386/*.c \
    contrib/asm386/*.bat contrib/asm386/zlibvc.d?? contrib/asm[56]86/*.?86 \
Only in d:\djgpp\contrib\zlib-1.1.4: Makefile.riscos
Only in d:\djgpp\contrib\zlib-1.1.4: maketree.c
Only in d:\djgpp\contrib\zlib-1.1.4: minigzip.c
Only in d:\djgpp\contrib\zlib-1.1.4: minigzip.d
Only in d:\djgpp\contrib\zlib-1.1.4: minigzip.exe
Only in d:\djgpp\contrib\zlib-1.1.4: msdos
Only in d:\djgpp\contrib\zlib-1.1.4: nt
Only in d:\djgpp\contrib\zlib-1.1.4: os2
diff -cwr ./README d:\djgpp\contrib\zlib-1.1.4/README
*** ./README	Sat Jul 08 17:51:40 2000
--- d:\djgpp\contrib\zlib-1.1.4/README	Mon Mar 11 10:02:54 2002
***************
*** 1,7 ****
! zlib 1.1.3 is a general purpose data compression library.  All the code
  is thread safe.  The data format used by the zlib library
  is described by RFCs (Request for Comments) 1950 to 1952 in the files 
! ftp://ds.internic.net/rfc/rfc1950.txt (zlib format), rfc1951.txt (deflate
  format) and rfc1952.txt (gzip format). These documents are also available in
  other formats from ftp://ftp.uu.net/graphics/png/documents/zlib/zdoc-index.html
  
--- 1,7 ----
! zlib 1.1.4 is a general purpose data compression library.  All the code
  is thread safe.  The data format used by the zlib library
  is described by RFCs (Request for Comments) 1950 to 1952 in the files 
! http://www.ietf.org/rfc/rfc1950.txt (zlib format), rfc1951.txt (deflate
  format) and rfc1952.txt (gzip format). These documents are also available in
  other formats from ftp://ftp.uu.net/graphics/png/documents/zlib/zdoc-index.html
  
***************
*** 14,64 ****
  
  To compile all files and run the test program, follow the instructions
  given at the top of Makefile. In short "make test; make install"
! should work for most machines. For Unix: "configure; make test; make install"
  For MSDOS, use one of the special makefiles such as Makefile.msc.
  For VMS, use Make_vms.com or descrip.mms.
  
! Questions about zlib should be sent to <zlib@quest.jpl.nasa.gov>, or to
  Gilles Vollant <info@winimage.com> for the Windows DLL version.
! The zlib home page is http://www.cdrom.com/pub/infozip/zlib/
! The official zlib ftp site is ftp://ftp.cdrom.com/pub/infozip/zlib/
! Before reporting a problem, please check those sites to verify that
  you have the latest version of zlib; otherwise get the latest version and
  check whether the problem still exists or not.
  
! Mark Nelson <markn@tiny.com> wrote an article about zlib for the Jan. 1997
  issue of  Dr. Dobb's Journal; a copy of the article is available in
! http://web2.airmail.net/markn/articles/zlibtool/zlibtool.htm
  
! The changes made in version 1.1.3 are documented in the file ChangeLog.
! The main changes since 1.1.2 are:
  
! - fix "an inflate input buffer bug that shows up on rare but persistent
!   occasions" (Mark)
! - fix gzread and gztell for concatenated .gz files (Didier Le Botlan)
! - fix gzseek(..., SEEK_SET) in write mode
! - fix crc check after a gzeek (Frank Faubert)
! - fix miniunzip when the last entry in a zip file is itself a zip file
!   (J Lillge)
! - add contrib/asm586 and contrib/asm686 (Brian Raiter)
!   See http://www.muppetlabs.com/~breadbox/software/assembly.html
! - add support for Delphi 3 in contrib/delphi (Bob Dellaca)
! - add support for C++Builder 3 and Delphi 3 in contrib/delphi2 (Davide Moretti)
! - do not exit prematurely in untgz if 0 at start of block (Magnus Holmgren)
! - use macro EXTERN instead of extern to support DLL for BeOS (Sander Stoks)
! - added a FAQ file
  
- plus many changes for portability.
  
  Unsupported third party contributions are provided in directory "contrib".
  
! A Java implementation of zlib is available in the Java Development Kit 1.1
  http://www.javasoft.com/products/JDK/1.1/docs/api/Package-java.util.zip.html
! See the zlib home page http://www.cdrom.com/pub/infozip/zlib/ for details.
  
  A Perl interface to zlib written by Paul Marquess <pmarquess@bfsec.bt.co.uk>
! is in the CPAN (Comprehensive Perl Archive Network) sites, such as:
! ftp://ftp.cis.ufl.edu/pub/perl/CPAN/modules/by-module/Compress/Compress-Zlib*
  
  A Python interface to zlib written by A.M. Kuchling <amk@magnet.com>
  is available in Python 1.5 and later versions, see
--- 14,63 ----
  
  To compile all files and run the test program, follow the instructions
  given at the top of Makefile. In short "make test; make install"
! should work for most machines. For Unix: "./configure; make test; make install"
  For MSDOS, use one of the special makefiles such as Makefile.msc.
  For VMS, use Make_vms.com or descrip.mms.
  
! Questions about zlib should be sent to <zlib@gzip.org>, or to
  Gilles Vollant <info@winimage.com> for the Windows DLL version.
! The zlib home page is http://www.zlib.org or http://www.gzip.org/zlib/
! Before reporting a problem, please check this site to verify that
  you have the latest version of zlib; otherwise get the latest version and
  check whether the problem still exists or not.
  
! PLEASE read the zlib FAQ http://www.gzip.org/zlib/zlib_faq.html
! before asking for help.
! 
! Mark Nelson <markn@ieee.org> wrote an article about zlib for the Jan. 1997
  issue of  Dr. Dobb's Journal; a copy of the article is available in
! http://dogma.net/markn/articles/zlibtool/zlibtool.htm
  
! The changes made in version 1.1.4 are documented in the file ChangeLog.
! The only changes made since 1.1.3 are bug corrections:
  
! - ZFREE was repeated on same allocation on some error conditions.
!   This creates a security problem described in
!   http://www.zlib.org/advisory-2002-03-11.txt
! - Returned incorrect error (Z_MEM_ERROR) on some invalid data
! - Avoid accesses before window for invalid distances with inflate window
!   less than 32K.
! - force windowBits > 8 to avoid a bug in the encoder for a window size
!   of 256 bytes. (A complete fix will be available in 1.1.5).
! 
! The beta version 1.1.5beta includes many more changes. A new official
! version 1.1.5 will be released as soon as extensive testing has been
! completed on it.
  
  
  Unsupported third party contributions are provided in directory "contrib".
  
! A Java implementation of zlib is available in the Java Development Kit
  http://www.javasoft.com/products/JDK/1.1/docs/api/Package-java.util.zip.html
! See the zlib home page http://www.zlib.org for details.
  
  A Perl interface to zlib written by Paul Marquess <pmarquess@bfsec.bt.co.uk>
! is in the CPAN (Comprehensive Perl Archive Network) sites
! http://www.cpan.org/modules/by-module/Compress/
  
  A Python interface to zlib written by A.M. Kuchling <amk@magnet.com>
  is available in Python 1.5 and later versions, see
***************
*** 117,123 ****
  
  Copyright notice:
  
!  (C) 1995-1998 Jean-loup Gailly and Mark Adler
  
    This software is provided 'as-is', without any express or implied
    warranty.  In no event will the authors be held liable for any damages
--- 116,122 ----
  
  Copyright notice:
  
!  (C) 1995-2002 Jean-loup Gailly and Mark Adler
  
    This software is provided 'as-is', without any express or implied
    warranty.  In no event will the authors be held liable for any damages
Only in .: Release
diff -cwr ./trees.c d:\djgpp\contrib\zlib-1.1.4/trees.c
*** ./trees.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/trees.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* trees.c -- output deflated data using Huffman coding
!  * Copyright (C) 1995-1998 Jean-loup Gailly
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* trees.c -- output deflated data using Huffman coding
!  * Copyright (C) 1995-2002 Jean-loup Gailly
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: trees.d
Only in d:\djgpp\contrib\zlib-1.1.4: trees.o
diff -cwr ./uncompr.c d:\djgpp\contrib\zlib-1.1.4/uncompr.c
*** ./uncompr.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/uncompr.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* uncompr.c -- decompress a memory buffer
!  * Copyright (C) 1995-1998 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* uncompr.c -- decompress a memory buffer
!  * Copyright (C) 1995-2002 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: uncompr.d
Only in d:\djgpp\contrib\zlib-1.1.4: uncompr.o
diff -cwr ./zconf.h d:\djgpp\contrib\zlib-1.1.4/zconf.h
*** ./zconf.h	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/zconf.h	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* zconf.h -- configuration of the zlib compression library
!  * Copyright (C) 1995-1998 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* zconf.h -- configuration of the zlib compression library
!  * Copyright (C) 1995-2002 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
***************
*** 186,193 ****
  #      endif
  #    endif
  #  endif
- #elif defined(_WIN32)
- #  define ZEXPORTVA __cdecl
  #endif
  
  #if defined (__BEOS__)
--- 186,191 ----
Only in d:\djgpp\contrib\zlib-1.1.4: zlib.3
Only in .: zlib.dsp
diff -cwr ./zlib.h d:\djgpp\contrib\zlib-1.1.4/zlib.h
*** ./zlib.h	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/zlib.h	Mon Mar 11 08:56:38 2002
***************
*** 1,7 ****
  /* zlib.h -- interface of the 'zlib' general purpose compression library
!   version 1.1.3, July 9th, 1998
  
!   Copyright (C) 1995-1998 Jean-loup Gailly and Mark Adler
  
    This software is provided 'as-is', without any express or implied
    warranty.  In no event will the authors be held liable for any damages
--- 1,7 ----
  /* zlib.h -- interface of the 'zlib' general purpose compression library
!   version 1.1.4, March 11th, 2002
  
!   Copyright (C) 1995-2002 Jean-loup Gailly and Mark Adler
  
    This software is provided 'as-is', without any express or implied
    warranty.  In no event will the authors be held liable for any damages
***************
*** 37,43 ****
  extern "C" {
  #endif
  
! #define ZLIB_VERSION "1.1.3"
  
  /* 
       The 'zlib' compression library provides in-memory compression and
--- 37,43 ----
  extern "C" {
  #endif
  
! #define ZLIB_VERSION "1.1.4"
  
  /* 
       The 'zlib' compression library provides in-memory compression and
Only in d:\djgpp\contrib\zlib-1.1.4: zlib.html
Only in .: zlib.plg
Only in .: zlib.vcproj
diff -cwr ./zutil.c d:\djgpp\contrib\zlib-1.1.4/zutil.c
*** ./zutil.c	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/zutil.c	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* zutil.c -- target dependent utility functions for the compression library
!  * Copyright (C) 1995-1998 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
--- 1,5 ----
  /* zutil.c -- target dependent utility functions for the compression library
!  * Copyright (C) 1995-2002 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h 
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: zutil.d
diff -cwr ./zutil.h d:\djgpp\contrib\zlib-1.1.4/zutil.h
*** ./zutil.h	Sat Jul 08 17:51:42 2000
--- d:\djgpp\contrib\zlib-1.1.4/zutil.h	Mon Mar 11 08:16:00 2002
***************
*** 1,5 ****
  /* zutil.h -- internal interface and configuration of the compression library
!  * Copyright (C) 1995-1998 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h
   */
  
--- 1,5 ----
  /* zutil.h -- internal interface and configuration of the compression library
!  * Copyright (C) 1995-2002 Jean-loup Gailly.
   * For conditions of distribution and use, see copyright notice in zlib.h
   */
  
Only in d:\djgpp\contrib\zlib-1.1.4: zutil.o
